bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+package org.mozartoz.bootcompiler
+package transform
+
+import ast._
+import oz._
+import symtab._
+import scala.collection._
+import com.sun.javafx.fxml.expression.BinaryExpression
+
+sealed class VarState(val onStack: Boolean, val locked: Boolean) {
+}
+object Declared      extends VarState(false, false)
+object Captured      extends VarState(false, false)
+object BoundFirst    extends VarState(true, true)
+object ToInstantiate extends VarState(false, true)
+
+case class VarAndState(v: Variable, state: VarState)
+
+/**
+ * Looks for declarations and the usages of those declarations.
+ * The Transform methods detect local statements and add declared variables to the environment
+ * The Walk methods are collecting their usages and tagging them
+ * 
+ * If a variable is bound after its declaration, we can shortcut it
+ * 
+ * When a proc is declared, it is difficult to track where it will end up. So if a call occurs,
+ * by precaution, all captured variable that were not bound should be instantiated
```

Comment:
Ok! It seemed indeed difficult to be very sure the proc would not end up in a particular variable. There may be some cases in which we could improve without taking any risk though. For instance if we tag the variable bound to the proc and we ensure it has not been used before the next call.

But then, we also have to check what variables the call really uses, which may again not be that obvious, ... This would require some more analysis for every call node and did not feel like worth it, referring to how the dynamic version of this feature performed.

However, this might be a nice improvement, but I would not consider it as minor. I think using a concept of "dependency tree" would definitely be a way to go.